pkg/kubelet/rkt/rkt.go

```diff
@@ -0,0 +1,750 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rkt
+
+import (
+	"encoding/json"
+	"fmt"
+	"hash/adler32"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kubecontainer "github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/container"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume"
+	"github.com/coreos/go-systemd/dbus"
+	"github.com/coreos/go-systemd/unit"
+	schema "github.com/coreos/rkt/Godeps/_workspace/src/github.com/appc/spec/schema"
+	appctypes "github.com/coreos/rkt/Godeps/_workspace/src/github.com/appc/spec/schema/types" // TODO(yifan): Fix this.
+	"github.com/coreos/rkt/cas"
+	"github.com/golang/glog"
+)
+
+const (
+	acversion            = "0.5.1"
+	kubernetesUnitPrefix = "k8s"
+	systemdServiceDir    = "/run/systemd/system"
+	tmpPodManifestDir    = "/tmp"
+
+	unitKubernetesSection = "X-Kubernetes"
+	unitPodName           = "POD"
+	unitRktID             = "RktID"
+
+	// TODO(yifan): Export in rkt?
+	defaultDataDir = "/var/lib/rkt"
+)
+
+const (
+	rktBinName = "rkt"
+
+	// TODO(yifan): Replace theses.
+	podStateEmbryo    = "embryo"
+	podStatePreparing = "preparing"
+	podStatePrepared  = "prepared"
+	podStateRunning   = "running"
+	podStateExited    = "exited"
+	podStateDeleting  = "deleting"
+	podStateGone      = "gone"
+)
+
+// Runtime implements the ContainerRuntime for rkt. The implementation
+// uses systemd, so in order to run this runtime, systemd must be installed
+// on the machine.
+type Runtime struct {
+	systemd *dbus.Conn
+	absPath string
+	config  *Config
+}
+
+// Config stores the global configuration for the rkt runtime.
+// Run 'rkt' for more details.
+type Config struct {
+	// The debug flag for rkt.
+	Debug bool
+	// The rkt data directory
+	Dir string
+	// This flag controls whether we skip image or key verification.
+	InsecureSkipVerify bool
+}
+
+// buildGlobalOptions returns an array of global command line options.
+func (c *Config) buildGlobalOptions() []string {
+	var result []string
+	if c == nil {
+		return result
+	}
+
+	result = append(result, fmt.Sprintf("--debug=%v", c.Debug))
+	result = append(result, fmt.Sprintf("--insecure-skip-verify=%v", c.InsecureSkipVerify))
+	if c.Dir != "" {
+		result = append(result, fmt.Sprintf("--dir=%s", c.Dir))
+	}
+	return result
+}
+
+// New creates the rkt container runtime which implements the container runtime interface.
```

Comment:
Can we check for the existence of systemd? Some way for Rkt to say, yes I can run on this machine